### DISCUSSION TITLE
Is it currently possible to show the tray icon on linux?

### DISCUSSION THREAD

**Discussion by rp1231** [houmain/keymapper#126] [Q&A] [answered]
I'm using the cinnamon desktop environment and the tray icon doesn't show up.
Would like it if the tray icon was present to indicate that keymapper is running and quickly access the conf file just like the windows version.

**Comment by ristomatti** [comment]
AFAIK there isn't a tray icon for Linux, so that's likely why you aren't seeing it. Create a launcher to edit the config with your favorite editor to solve quick access to the file? E.g. something like this:

```ini
[Desktop Entry]
Icon=/some/path.svg
Exec=vim ~/.config/keymapper.conf
Version=1.0.0
Type=Application
Categories=Utility;
Name=Edit Keymapper config
Terminal=true
```

...goes into `~/.local/share/applications/edit-keymamaper-config.desktop`

**Comment by houmain** [comment]
Yes, right, there is only an icon on Windows. But I think I can add one on Linux using libappindicator.

**Reply by ristomatti** [reply]
Consider putting this behind a flag or even a separate command (think NetworkManager's `nm-applet`). The reasoning here is that Keymapper can also function without Xorg/Wayland running. The flag/binary could be included in `keymapper.desktop` as a default.

A separate binary would create a basis to later add other features besides editing/reloading the config or exiting `keymapper`, like stop/start action. This could of course just be a toggle switch to temporarily pause remapping. A stopped/paused state could be indicated with a greyed out try icon.

**Reply by rp1231** [reply]
It would be great if you could implement it.
Thanks

**Reply by houmain** [reply]
@ristomatti thanks for the food for thought. I think I will implement it like in the Windows version for now: enable it by default and allow to disable it using the `--no-tray` argument.
It will be possible to compile `keymapper` without it and when it is compiled in, it will be initialized only when available and still work when it is not (like the X11/Wayland/D-Bus focused window detection and string typing).

**Reply by ristomatti** [reply]
Consistency between the versions is not bad either.

**Reply by houmain** [reply]
Hi, the tray icon is testable now in the 4.1.0 release.

**Reply by ristomatti** [reply]
I quickly tried it yesterday. It appeared to work but the menu seems to take comparatively long to appear. I would also consider showing the menu only by right clicking. Left click could either toggle or do nothing. Only two other apps show the menu from either click:

```
Barrier     left click -> n/a,              right click -> show menu
DeskDock    left click -> n/a,              right click -> show menu

Solaar      left click -> show menu,        right click -> show menu
JB Toolbox  left click -> show menu,        right click -> show menu

Flameshot   left click -> take screenshot,  right click -> show menu
NetworkMgr. left click -> connections menu, right click -> show menu

Slack       left click -> show app window,  right click -> show menu
Telegram    left click -> show app window,  right click -> show menu
Teams2Linux left click -> show app window,  right click -> show menu
```

**Reply by rp1231** [reply]
I installed it from the aur and it seems to work for me as well, but I too am experiencing the slowness of the menu.
Also the icon doesn't seem to show up on manjaro cinnamon edition. I get a placeholder tray icon instead of the keymapper tray icon.
![Screenshot from 2024-04-18 19-14-43](https://github.com/houmain/keymapper/assets/47380919/e2640fd3-7af5-4a81-9e4d-e5543d669565)
As you can see here, the keymapper icon is the one to the left of the red shield icon.

**Reply by rp1231** [reply]
This is the console ouput when I ran keymapper from the terminal:
```
(keymapper:12192): Gdk-CRITICAL **: 19:13:18.548: gdk_window_thaw_toplevel_updates: assertion 'window->update_and_descendants_freeze_count > 0' failed

(xed:12274): libenchant-WARNING **: 19:13:38.614: Error loading plugin: libaspell.so.15: cannot open shared object file: No such file or directory


(xed:12274): libenchant-WARNING **: 19:13:38.614: Error loading plugin: libhspell.so.0: cannot open shared object file: No such file or directory


(xed:12274): libenchant-WARNING **: 19:13:38.615: Error loading plugin: libnuspell.so.5: cannot open shared object file: No such file or directory


(xed:12274): libenchant-WARNING **: 19:13:38.615: Error loading plugin: libvoikko.so.1: cannot open shared object file: No such file or directory


(xed:12274): libenchant-WARNING **: 19:13:38.615: Error loading plugin: libaspell.so.15: cannot open shared object file: No such file or directory


(xed:12274): libenchant-WARNING **: 19:13:38.615: Error loading plugin: libhspell.so.0: cannot open shared object file: No such file or directory


(xed:12274): libenchant-WARNING **: 19:13:38.615: Error loading plugin: libnuspell.so.5: cannot open shared object file: No such file or directory


(xed:12274): libenchant-WARNING **: 19:13:38.615: Error loading plugin: libvoikko.so.1: cannot open shared object file: No such file or directory

Version 4.1.0.r0.g472b4ca
(c) 2019-2024 by Albert Kalchmair

All Rights Reserved.
This program comes with absolutely no warranty.
See the GNU General Public License, version 3 for details.
Configuration updated

```

**Reply by rp1231** [reply]
But when I run keymapper on arch-gnome, the tray icon runs perfectly and there's no lag in the menu as well.
Even the icon shows up.

**Reply by ristomatti** [reply]
The warnings above give the impression there's some issues on the distro level. There's no chance Keymapper has anything to do with Finnish language spelling (`libvoikko`) for instance. Possibly [this issue](https://forum.manjaro.org/t/gedit-plugin-warning-when-opening-from-terminal/75437).

The very first line I see also on my Ubuntu 20.04 based system running Regolith DE (i3wm). I've reported it as an issue here #131.

**Reply by ristomatti** [reply]
I seem to get a placeholder icon as well:

![image](https://github.com/houmain/keymapper/assets/9029939/72dc6211-d109-4986-8588-49f7cb736c0e)

I'm assuming the icon should look something like this one included in the repo as [extra/icon.svg](https://github.com/houmain/keymapper/blob/main/extra/icon.svg):
 
![image](https://github.com/houmain/keymapper/assets/9029939/1a8f48c7-d4fd-4dc5-8a70-f298a0343eab)

**Reply by houmain** [reply]
No, currently it should look like a keyboard-indicator.
![image](https://github.com/houmain/keymapper/assets/20393908/1d7dd8f4-192e-467c-9f27-0fa55dec99f7)
That is also how the icon is called "indicator-keyboard". I hoped that this was commonly available, but I did not look carefully enough. All the versions I have are part of the Papirus theme I installed.

**Reply by ristomatti** [reply]
@houmain To illustrate the issue:

https://github.com/houmain/keymapper/assets/9029939/713affe5-ff66-45a3-8a92-e39cd4353554

(yes, I accidentally closed Flameshot in the process)

**Reply by ristomatti** [reply]
That's what I was thinking as well when I glanced over the changes. It seems like it might depend on the selected icon set. It appears to be common indeed:

```
$ fd indicator-keyboard /usr/share/icons | wc -l
1716
```


...but maybe just not on the one I picked based on the general look. I'll see if I get it to show. I pretty much see the icons when I occasionally open Nemo anyway.

**Reply by houmain** [reply]
Thanks, yes really strange that delay. Hopefully it disappears when it can find the icon. I just made a 4.1.1 release to fix that.

**Reply by ristomatti** [reply]
Yup, depends on the selected icon theme. It was included with "Arc", "Gnome", Epapirus" and a few other but surprisingly not with Linux Mint's default "Adwaita", "Mint-X", "Pop" or "ubuntu-mono-dark". I'll try just copying the best looking variant over to "Newaita-dark" I'm using and see if it does the trick. :slightly_smiling_face:

**Reply by ristomatti** [reply]
What, you released a new version during the time I was going through the icons. 🤯

**Reply by ristomatti** [reply]
Yea the Papirus version looked great, Gnome not so much. Papirus Light was too dim for my dark grey theme while Papirus Dark was blinding white. :joy:

With the update, I now get the "Key" icon. A bit blurry with my minimalist desktop but it's much better, e.g. doesn't look like a missing icon. Good enough for me. :+1:

![image](https://github.com/houmain/keymapper/assets/9029939/9f73db63-40cc-404f-9431-d5c9f1c13980)

...but the lag is still there. It's not an issue for me as for me it's an indicator the process is running and can also act as an "emergency" exit mechanism if I manage to mess up the bindings in a way I can't save them or open the browser to check what the exit key was. This has happened once and I had to kill the process via ssh from my phone as I had a lot of stuff open. :joy:

**Reply by ristomatti** [reply]
To edit the config, I hit `Space+BracketLeft` to pull out a Tmux scratchpad which usually has my keymapper window already open. :grin: 


https://github.com/houmain/keymapper/assets/9029939/5f59fddb-61f4-4fbb-ae22-4e05205eed02

**Reply by ristomatti** [reply]
BTW now that I checked the `journalctl` log, this issue was clearly visible:

![image](https://github.com/houmain/keymapper/assets/9029939/5e34a365-db55-45a9-ae2f-6521cc40cf3b)

**Reply by houmain** [reply]
I cannot reproduce the slowness of the menu but I think I know how to solve it. 
@ristomatti maybe you could try it first?
I think a for loop in "src/client/unix/TrayIconGtk.cpp" line 76 like the following could be enough:
```
  void update() override {
    for (auto i = 0; i < 3; ++i)
      gtk_main_iteration_do(false);
  }
```
Thanks!

**Reply by ristomatti** [reply]
Magic numbers. That did indeed improve it but it was still a bit laggy. I experimented with the iteration count 3, 5, 7 and 10 as well. Each step improved the responsiveness. 7 was almost unnoticeable but at 10 I can't spot the difference and it's a nice round number. 

I'd be curious to know what it does and why it helps.

**Reply by houmain** [reply]
Thank you very much! I am glad I asked you to try it first. I misinterpreted the documentation. I thought it was processing all pending events. The correct solution is:
```
while (gtk_events_pending())
  gtk_main_iteration();
```

**Reply by ristomatti** [reply]
This makes a lot more sense.

**Reply by ristomatti** [reply]
Confirm. Now it appears instantly.

**Reply by rp1231** [reply]
Can confirm that it works for me instantly as well on fedora cinnamon.
Even the icon appears and there is no lag whatsoever.
But when I click on the configuration option, I get this message:
```
gio: file:///home/fedora/.config/keymapper.conf: Failed to find default application for content type ‘application/x-zerosize’
```
It should open the .conf file in the default text editor right?

**Reply by ristomatti** [reply]
@rp1231 Yep. Is the config file on that location empty or missing as the content type is `application/x-zerosize`? Do you get the same error with `xdg-open 'file:///home/fedora/.config/keymapper.conf'`? If I run it, the config opens in the default editor.

**Reply by rp1231** [reply]
@ristomatti  Yes the config file was empty at that point. The problem was that .conf files didn't have any default application set for them. Once I set the text editor as the default program, the problem got solved.
Thanks for your help.

**Reply by houmain** [reply, marked as answer]
With the 4.1.3 release it should create a configuration file (containing just a single newline, so the MIME type can be detected correctly) and open it in the text editor.

**Comment by ristomatti** [comment]
@houmain BTW the "Devices" list as a notification would not be of much use to me as Rofication shows it as a single line of text and I only see ~4 devices until it truncates the line.
